app_integrations/apps/box.py

```diff
@@ -0,0 +1,211 @@
+            https://developer.box.com/reference#section-date-format
+        """
+        return '%Y-%m-%dT%H:%M:%S-00:00'
+
+    @classmethod
+    def _load_auth(cls, auth_data):
+        """Load JWTAuth from Box service account JSON keyfile
+
+        Args:
+            auth_data (dict): The loaded keyfile data from a Box service account
+                JSON file
+        Returns:
+            boxsdk.JWTAuth Instance of JWTAuth that allows the client to authenticate
+                or False if there was an issue loading the auth
+        """
+        try:
+            auth = JWTAuth.from_settings_dictionary(auth_data)
+        except (TypeError, ValueError, KeyError):
+            LOGGER.exception('Could not load JWT from settings dictionary')
+            return False
+
+        return auth
+
+    def _create_client(self):
+        """Box requests must be signed with a JWT keyfile
+
+        Returns:
+            bool: True if the Box client was successfully created or False
+                if any errors occurred during the creation of the client
+        """
+        if self._client:
+            LOGGER.debug('Client already instantiated for %s', self.type())
+            return True
+
+        auth = self._load_auth(self._config.auth)
+        if not auth:
+            return False
+
+        self._client = Client(auth)
+
+        return bool(self._client)
+
+    def _make_request(self):
+        """Make the request using the Box client
+
+        The inner function of `_perform_request` is used to handle a single retry in
+        the event of a ConnectionError. If this fails twice, the function will return
+
+        Returns:
+            dict: Response from Box (boxsdk.session.box_session.BoxResponse) that is
+                json loaded into a dictionary.
+        """
+        # Create the parameters for this request, 100 is the max value for limit
+        params = {
+            'limit': self._MAX_CHUNK_SIZE,
+            'stream_type': EnterpriseEventsStreamType.ADMIN_LOGS,
+        }
+
+        # From Box's docs: Box responds to the created_before and created_after
+        # parameters only if the stream_position parameter is not included.
+        if self._next_stream_position:
+            params['stream_position'] = self._next_stream_position
+        else:
+            params['created_after'] = self._last_timestamp
+
+        def _perform_request(allow_retry=True):
+            try:
+                # Get the events using a make_request call with the box api. This is to
+                # support custom parameters such as 'created_after' and 'created_before'
+                box_response = self._client.make_request(
+                    'GET',
+                    self._client.get_url('events'),
+                    params=params
+                )
+            except BoxException:
+                LOGGER.exception('Failed to get events for %s', self.type())
+                return False
+            except ConnectionError:
```

Comment:
Did you consider using the `backoff` library for this? It's designed for this exact use case

Reply:
I had, yes, but this isn't an error that's caused by too many requests, etc and is an error that is caused (I think) by a bug in the box sdk where it is not properly refreshing the auth token (again, I think). Here is a sample of the traceback I saw once or twice during testing:
```
Traceback (most recent call last):
  File "/Users/my_user/Desktop/testopen.py", line 205, in gather_logs
    params=params
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/util/api_call_decorator.py", line 71, in call
    return method(*args, **kwargs)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/client/client.py", line 407, in make_request
    return self._session.request(method, url, **kwargs)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/session/box_session.py", line 510, in request
    response = self._prepare_and_send_request(method, url, **kwargs)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/session/box_session.py", line 350, in _prepare_and_send_request
    **kwargs
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/session/box_session.py", line 395, in _make_request
    access_token_will_be_used = self._renew_session(None)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/session/box_session.py", line 196, in _renew_session
    new_access_token, _ = self._oauth.refresh(access_token_used)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/oauth2.py", line 251, in refresh
    access_token, refresh_token = self._refresh(access_token_to_refresh)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/jwt_auth.py", line 394, in _refresh
    new_access_token = self.authenticate_instance()
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/jwt_auth.py", line 384, in authenticate_instance
    return self._auth_with_jwt(self._enterprise_id, 'enterprise')
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/jwt_auth.py", line 243, in _auth_with_jwt
    return self._construct_and_send_jwt_auth(sub, sub_type)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/jwt_auth.py", line 219, in _construct_and_send_jwt_auth
    return self.send_token_request(data, access_token=None, expect_refresh_token=False)[0]
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/oauth2.py", line 359, in send_token_request
    token_response = self._execute_token_request(data, access_token, expect_refresh_token)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/auth/oauth2.py", line 328, in _execute_token_request
    access_token=access_token,
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/boxsdk/network/default_network.py", line 24, in request
    request_response=self._session.request(method, url, **kwargs),
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/my_venv/stream_alert_venv/lib/python2.7/site-packages/requests/adapters.py", line 490, in send
    raise ConnectionError(err, request=request)
ConnectionError: ('Connection aborted.', error(54, 'Connection reset by peer'))
```

I will probably implement backoff in the apps at some point, but will want to do it on broader level and not a per-app one off.